### PR TITLE
Tabs block update/enhancements

### PR DESCRIPTION
**Pull request by ryanmparrish**
1. Update tabs to use milo:defered to stop content/image flicker - tabs 2 and 3 with icons.
2. Update tab content spacing so tab content background color is seamless inside tab content. Currently there is a gap above and below content area
3. The tab content section-metadata style property...
- "grid width" (grid width 6-10) pushes content off screen. See tab 4 in sample url below.
- "three up" (2-5 ups) isn't respected. See tab 1.
4. Update tab reference in section-metadata to identify the tab via `tab-list ${index}` instead of by `label` (1, 2 ,3, etc.) to pair the section to the correct tab. This is largely to better support localization. 
- NOTE: For the tab by index relation, It is necessary to include a `tabs` block `id` table row with a unique `value`

tab-block-by-id-authoring

* Backward compatible with what's out in the wild using the current tab label reference pattern.

Resolves: [MWPW-127373](https://jira.corp.adobe.com/browse/MWPW-127373)

**Test URLs:**
- [Before](https://main--milo--adobecom.hlx.page/drafts/rparrish/tabs-sample?martech=off)
- [After](https://rparrish-tabs-update--milo--adobecom.hlx.page/drafts/rparrish/tabs-sample?martech=off)

**Live Docs:**
- [Before: https://milo.adobe.com/docs/authoring/links](https://milo.adobe.com/docs/authoring/links)
- [After: https://rparrish-tabs-update--milo--adobecom.hlx.page/docs/authoring/links?milolibs=rparrish-tabs-update](https://rparrish-tabs-update--milo--adobecom.hlx.page/docs/authoring/links?milolibs=rparrish-tabs-update)

**Test Tabs in Modal**
- [Before: https://main--milo--adobecom.hlx.page/about/?akamaiLocale=CH](https://main--milo--adobecom.hlx.page/about/?akamaiLocale=CH)
- [After: https://rparrish-tabs-update--milo--adobecom.hlx.page/about/?akamaiLocale=CH](https://rparrish-tabs-update--milo--adobecom.hlx.page/about/?akamaiLocale=CH)

**Test BACOM**
- [Before: https://business.adobe.com/customer-success-stories/pga-tour-case-study.html](https://business.adobe.com/customer-success-stories/pga-tour-case-study.html)
- [After: https://business.adobe.com/customer-success-stories/pga-tour-case-study.html?milolibs=rparrish-tabs-update#tabs-0](https://business.adobe.com/customer-success-stories/pga-tour-case-study.html?milolibs=rparrish-tabs-update#tabs-0)

**Test DC**
- [Before: https://main--cc--adobecom.hlx.page/drafts/tkato/cc-overview-non-xf-0404-backup-most-recent](https://main--cc--adobecom.hlx.page/drafts/tkato/cc-overview-non-xf-0404-backup-most-recent)
- [After: https://main--cc--adobecom.hlx.page/drafts/tkato/cc-overview-non-xf-0404-backup-most-recent?milolibs=rparrish-tabs-update](https://main--cc--adobecom.hlx.page/drafts/tkato/cc-overview-non-xf-0404-backup-most-recent?milolibs=rparrish-tabs-update)

### CODE REVIEW

File: libs/blocks/tabs/tabs.js [hunk 1]
```diff
@@ -84,13 +87,41 @@ function initTabs(elm, config, rootElem) {
...
   if (!rows.length) return;
+  block.addEventListener('milo:deferred', handleDeferredImages(block), true);
+
+  // Tab Config
+  const config = {};
+  const configRows = [].slice.call(rows);
+  configRows.splice(0, 1);
+  if (configRows) {
+    configRows.forEach((row) => {
+      console.log('config row', row.children[0].textContent);
```
Comment (Sartxi) [inline libs/blocks/tabs/tabs.js:113]: Are these row config logs being used? Maybe they could be deleted?

**Review by Sartxi — APPROVED** [review]
Looks good to me!

File: libs/blocks/tabs/tabs.js (hunk 1 above)
Comment (ryanmparrish) [reply libs/blocks/tabs/tabs.js:113]: Good eye, No thats was testing some stuff. Removed now

File: libs/blocks/tabs/tabs.js [hunk 2]
```diff
@@ -84,13 +87,39 @@ function initTabs(elm, config, rootElem) {
...
   const rootElem = block.closest('.fragment') || document;
   const rows = block.querySelectorAll(':scope > div');
+  const parentSection = block.closest('.section');
   /* c8 ignore next */
   if (!rows.length) return;
+  block.addEventListener('milo:deferred', handleDeferredImages(block), true);
+
+  // Tab Config
+  const config = {};
+  const configRows = [].slice.call(rows);
```
Comment (meganthecoder) [inline libs/blocks/tabs/tabs.js:109]: I'm not familiar with this pattern. Is it the same as `[...rows]`?

File: libs/blocks/tabs/tabs.js [hunk 3]
```diff
@@ -84,13 +87,39 @@ function initTabs(elm, config, rootElem) {
...
+  configRows.splice(0, 1);
+  if (configRows) {
+    configRows.forEach((row) => {
+      const rowKey = getStringKeyName(row.children[0].textContent);
+      const rowVal = row.children[1].textContent.trim();
+      config[rowKey] = rowVal;
+      row.remove();
+    });
+  }
+  const tabsId = config.id ? config.id : getUniqueId(block, rootElem);
```
Comment (meganthecoder) [inline libs/blocks/tabs/tabs.js:119]: Nit
```suggestion
  const tabsId = config.id || getUniqueId(block, rootElem);
```

File: libs/blocks/tabs/tabs.js [hunk 4]
```diff
@@ -109,64 +136,57 @@ const init = (block) => {
...
-      .forEach((d) => {
-        const metaValue = getStringKeyName(d.children[1].textContent);
-        const section = sectionMetadata.closest('.section');
-        const assocTabItem = rootElem.querySelector(`#tab-panel-${initCount}-${CSS.escape(metaValue)}`);
-        if (assocTabItem) assocTabItem.append(section);
-      });
+    const rows = sectionMetadata.querySelectorAll(':scope > div');
+    rows.forEach((row) => {
+      const key = getStringKeyName(row.children[0].textContent);
+      if (key != 'tab') return;
```
Comment (meganthecoder) [inline libs/blocks/tabs/tabs.js:175]: Is there a reason you aren't using strict inequality?

File: libs/blocks/tabs/tabs.css [hunk 5]
```diff
@@ -1,3 +1,5 @@
+/* stylelint-disable selector-class-pattern */
```
[the remark this answers is not in the document]
Comment (ryanmparrish) [reply libs/blocks/tabs/tabs.css]: Agree. I removed this and updated all camelCase class names to be kebab-case. Thanks

File: libs/blocks/tabs/tabs.js (hunk 2 above)
Comment (ryanmparrish) [reply libs/blocks/tabs/tabs.js:109]: Both methods are the same and can be used based on personal preference. I might update this to use the spread operator to keep more consistent w/ other block methods. consistency is key.

File: libs/blocks/tabs/tabs.js (hunk 4 above)
Comment (ryanmparrish) [reply libs/blocks/tabs/tabs.js]: Derp, nope - just an oversight. This is updated, thanks!

**Review by meganthecoder — APPROVED** [review]

File: libs/features/georoutingv2/georoutingv2.css [hunk 6]
```diff
@@ -78,7 +78,7 @@
   background: initial;
 }
 
-.dialog-modal.locale-modal-v2 .tabs div[role="tablist"] .tabList-container {
+.dialog-modal.locale-modal-v2 .tabs div[role="tablist"] .tab-list-container {
```
Comment (honstar) [inline libs/features/georoutingv2/georoutingv2.css:81]: There are `georoutingv2` unit-test failures, not sure if this is related to these changes? Can you please check?

Comment (ryanmparrish) [reply libs/features/georoutingv2/georoutingv2.css:81]: Glad we have tests. I'm seeing an issue at https://rparrish-tabs-update--milo--adobecom.hlx.page/about/?akamaiLocale=CH
Looking into it...

Comment (ryanmparrish) [reply libs/features/georoutingv2/georoutingv2.css:81]: I figured out the issue and have pushed up a fix. The scope of the `rootElement` changes in a fragment and was throwing off a selector in the associated tab query. I added this use case in the desc.

Comment (honstar) [reply libs/features/georoutingv2/georoutingv2.css:81]: Thanks @ryanmparrish! However, the links in the description under `Test Tabs in Frags` doesn't use a document with fragments at all.

Comment (rgclayton) [reply libs/features/georoutingv2/georoutingv2.css:81]: Yes, geo routing launches a modal. Not a fragment embedded in the page per se. But, upon detection. 
https://github.com/adobecom/milo/blob/main/libs/features/georoutingv2/georoutingv2.js#L280

**Review by rgclayton — APPROVED** [review]

File: libs/blocks/tabs/tabs.js [hunk 7]
```diff
@@ -84,89 +87,109 @@ function initTabs(elm, config, rootElem) {
...
+  e.removeEventListener('milo:deferred', handleDeferredImages, true);
+}
 
 const init = (block) => {
   const rootElem = block.closest('.fragment') || document;
   const rows = block.querySelectorAll(':scope > div');
+  const parentSection = block.closest('.section');
   /* c8 ignore next */
   if (!rows.length) return;
+  block.addEventListener('milo:deferred', handleDeferredImages(block), true);
```
Comment (chrischrischris) [inline libs/blocks/tabs/tabs.js:105]: This runs `handleDeferredImages` right away.  Also the `milo:deferred` event is only fired on the `document`. and not each individual block.  Something like:

```suggestion
  handleDeferredImages(block);
```

```
const handleDeferredImages = (block) => {
  const loadLazyImages = () => {
    const images = block.querySelectorAll('img[loading="lazy"]');
		  images.forEach((img) => {
		    /* c8 ignore next */
		    img.removeAttribute('loading');
		  });
		  document.removeEventListener('milo:deferred', loadLazyImages, true);
  };
  document.addEventListener('milo:deferred', loadLazyImages, true);
}
```

Comment (chrischrischris) [reply libs/blocks/tabs/tabs.js:105]: Might be premature optimization, but any images on a tab that is not currently displayed can have the image loading left to lazy.

File: libs/blocks/tabs/tabs.js [hunk 8]
```diff
@@ -84,89 +87,109 @@ function initTabs(elm, config, rootElem) {
...
+
+  // Tab Config
+  const config = {};
+  const configRows = [...rows];
+  configRows.splice(0, 1);
+  if (configRows) {
```
Comment (chrischrischris) [inline libs/blocks/tabs/tabs.js:111]: No need for this check as `configRows` will always be an array

File: libs/blocks/tabs/tabs.js [hunk 9]
```diff
@@ -84,89 +87,109 @@ function initTabs(elm, config, rootElem) {
...
+    configRows.forEach((row) => {
+      const rowKey = getStringKeyName(row.children[0].textContent);
+      const rowVal = row.children[1].textContent.trim();
+      config[rowKey] = rowVal;
+      row.remove();
+    });
+  }
+  const tabsId = config.id || getUniqueId(block, rootElem);
```
Comment (chrischrischris) [inline libs/blocks/tabs/tabs.js:119]: nit: should this be `tabId`?

**Review by chrischrischris — APPROVED** [review]

**Review by chrischrischris — CHANGES_REQUESTED** [review]
Mistakenly approved earlier, requesting changes so that feedback is addressed.

File: libs/blocks/tabs/tabs.js (hunk 7 above)
Comment (ryanmparrish) [reply libs/blocks/tabs/tabs.js]: @chrischrischris  - Thanks for the feedback on this. I've addressed this file for now. 
LMK if we want to add a `milodefer` func to utils.

Comment (chrischrischris) [reply libs/blocks/tabs/tabs.js]: No need to add `milodefer` for now until we have more use cases.

File: libs/blocks/tabs/tabs.js [hunk 10]
```diff
@@ -84,89 +87,110 @@ function initTabs(elm, config, rootElem) {
...
+const handleDeferredImages = (block) => {
+  const loadLazyImages = () => {
+    const images = block.querySelectorAll('img[loading="lazy"]');
+		  images.forEach((img) => {
+		    /* c8 ignore next */
+		    img.removeAttribute('loading');
+		  });
+		  document.removeEventListener('milo:deferred', loadLazyImages, true);
+  };
+  document.addEventListener('milo:deferred', loadLazyImages, true);
```
Comment (chrischrischris) [inline libs/blocks/tabs/tabs.js:99]: Just realized we can probably use `once` so we don't have to `removeEventListener`.
```suggestion
  document.addEventListener('milo:deferred', loadLazyImages, { once: true, capture: true });
```

**Review by chrischrischris — APPROVED** [review]

**Pull request merged**